app/src/main/java/com/example/mvpmusicapp/data/source/local/LocalDataSource.kt

```diff
@@ -0,0 +1,46 @@
+package com.example.mvpmusicapp.data.source.local
+
+import android.content.ContentResolver
+import android.provider.MediaStore
+import android.util.Log
+import com.example.mvpmusicapp.data.model.Song
+import com.example.mvpmusicapp.data.source.SongDataSource
+
+class LocalDataSource private constructor(private val contentResolver: ContentResolver) :
+    SongDataSource.Local {
+    override fun getSongs(listener: OnDataLoadedCallback<MutableList<Song>>) {
+        LocalAsyncTask(listener) {
+            getSongFromDevice()
+        }.execute()
+    }
+
+    private fun getSongFromDevice(): MutableList<Song> {
+        val songs = mutableListOf<Song>()
+        val selection = MediaStore.Audio.Media.IS_MUSIC + " != 0"
+        val projection = arrayOf(
+            MediaStore.Audio.Media._ID,
+            MediaStore.Audio.Media.TITLE,
+            MediaStore.Audio.Media.ALBUM_ARTIST,
+            MediaStore.Audio.ArtistColumns.ARTIST
+        )
+        val cursor = contentResolver.query(
+            MediaStore.Audio.Media.EXTERNAL_CONTENT_URI,
+            projection,
+            selection,
+            null,
+            null
+        )
+        cursor?.let {
+            while (it.moveToNext()) {
+                songs.add(Song(it))
+            }
+        }
+        return songs
+    }
+
+    companion object {
+        private var instance: LocalDataSource? = null
+        fun getInstance(contentResolover: ContentResolver) =
+            instance ?: LocalDataSource(contentResolover).also { instance = it }
```

Comment:
câu lệnh này nghĩa là gì ? 
Thêm cái also vào có ý nghĩa gì ?

Reply:
nó là singleton, giúp tạo 1 đối tượng duy nhất, cái also là có sẽ có it = chính cái đối tượng gọi also

Reply:
câu lệnh also sẽ được thực hiện sau khi return hay trước khi return thế ?

Reply:
theo em nghĩ thì sẽ là sau , thì it mới có giá trị để dùng

Reply:
vậy em thử viết không dùng also xem như nào ?

Reply:
nếu không dùng also , câu lệnh em nghĩ sẽ như này 
fun getInstance(contentResolover: ContentResolver) = if(instance == null) LocalDataSource(contentResolover) else instance 
?: có nghĩa là nếu null thì